engine/src/main/java/org/pentaho/di/trans/Trans.java

```diff
@@ -0,0 +1,5804 @@
+
+          addTransListener( new TransAdapter() {
+            @Override
+            public void transFinished( Trans trans ) {
+              timer.cancel();
+            }
+          } );
+        }
+      } catch ( KettleException e ) {
+        throw new KettleTransException( BaseMessages.getString( PKG, "Trans.Exception.ErrorWritingLogRecordToTable",
+          logTable ), e );
+      } finally {
+        // If we use interval logging, we keep the connection open for performance reasons...
+        //
+        if ( transLogTableDatabaseConnection != null && ( intervalInSeconds <= 0 ) ) {
+          transLogTableDatabaseConnection.disconnect();
+          transLogTableDatabaseConnection = null;
+        }
+      }
+    } catch ( KettleException e ) {
+      throw new KettleTransException( BaseMessages.getString( PKG,
+        "Trans.Exception.UnableToBeginProcessingTransformation" ), e );
+    }
+  }
+
+  /**
+   * Writes log channel information to a channel logging table (if one has been configured).
+   *
+   * @throws KettleException if any errors occur during logging
+   */
+  protected void writeLogChannelInformation() throws KettleException {
+    Database db = null;
+    ChannelLogTable channelLogTable = transMeta.getChannelLogTable();
+
+    // PDI-7070: If parent trans or job has the same channel logging info, don't duplicate log entries
+    Trans t = getParentTrans();
+    if ( t != null ) {
+      if ( channelLogTable.equals( t.getTransMeta().getChannelLogTable() ) ) {
+        return;
+      }
+    }
+
+    Job j = getParentJob();
+
+    if ( j != null ) {
+      if ( channelLogTable.equals( j.getJobMeta().getChannelLogTable() ) ) {
+        return;
+      }
+    }
+    // end PDI-7070
+
+    try {
+      db = new Database( this, channelLogTable.getDatabaseMeta() );
+      db.shareVariablesWith( this );
+      db.connect();
+      db.setCommit( logCommitSize );
+
+      List<LoggingHierarchy> loggingHierarchyList = getLoggingHierarchy();
+      for ( LoggingHierarchy loggingHierarchy : loggingHierarchyList ) {
+        db.writeLogRecord( channelLogTable, LogStatus.START, loggingHierarchy, null );
+      }
+
+      // Also time-out the log records in here...
+      //
+      db.cleanupLogRecords( channelLogTable, getName() );
+    } catch ( Exception e ) {
+      throw new KettleException( BaseMessages.getString( PKG,
+        "Trans.Exception.UnableToWriteLogChannelInformationToLogTable" ), e );
+    } finally {
+      disconnectDb( db );
+    }
+  }
+
+  /**
+   * Writes step information to a step logging table (if one has been configured).
+   *
+   * @throws KettleException if any errors occur during logging
+   */
+  protected void writeStepLogInformation() throws KettleException {
+    Database db = null;
+    StepLogTable stepLogTable = getTransMeta().getStepLogTable();
+    try {
+      db = createDataBase( stepLogTable.getDatabaseMeta() );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *db*. The resource is referenced at lines 2625, 2626, 2627, 2630, 2633, 2638. The resource is closed at line 2722. There are other execution paths that do not contain closure statements, for example, when Database.commit (in method disconnectDb) throws an exception, and an if condition at line 2719 evaluates to true.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).